javascript/organizing_your_javascript_code/organizing_code_with_objects.md

```diff
@@ -0,0 +1,374 @@
+
+A rock paper scissors game might involve a couple basic things:
+
+- Players' scores
+- The ability to play a round (and playing a round should update a player's score)
+
+And might also include a couple nice-to-haves
+
+- The ability to determine the current winning player
+- The ability to restart the game
+
+So, at it's most basic, an object that represents the game might look something like this (assuming we're playing against a computer player):
+
+```javascript
+const rps = {
+  playerScore: 0,
+  computerScore: 0,
+  playRound(playerChoice) {
+    // code to play the round... (and update the scores when a player wins)
+  }
+}
+```
+
+And if we fleshed it out, our object may come out to look something like this:
+
+```javascript
+const rps = {
+  playerScore: 0,
+  computerScore: 0,
+  playRound(playerChoice) {
+    const options = ['rock', 'paper', 'scissors']
+
+    // if an invalid choice is chosen, throw an error
+    if(!options.includes(playerChoice.toLowerCase())) {
+      throw new Error(`Expected 'rock', 'paper', or 'scissors', but got ${playerChoice}`)
+    }
+    
+    // get the computer's choice
+    const computerChoice = options[Math.floor(Math.random() * 3)]
+
+
+    // determine the winner, apply points if necessary, and return who won
+    if(playerChoice.toLowerCase() === computerChoice) {
+      return "tie"
+    } else if(
+      (playerChoice === 'rock' && computerChoice === 'scissors') ||
+      (playerChoice === 'paper' && computerChoice === 'rock') ||
+      (playerChoice === 'scissors' && computerChoice === 'paper')
+    ) {
+      this.playerScore++
+      return "player"
+    } else {
+      this.computerScore++
+      return 'computer'
+    }
+  },
+  getWinningPlayer() {
+    if(this.playerScore === this.computerScore) {
+      return 'tie'
+    } else if (this.playerScore > this.computerScore) {
+      return 'player'
+    } else {
+      return 'computer'
+    }
+  },
+  reset() {
+    this.playerScore = 0;
+    this.computerScore = 0;
+  }
+}
+
+rps.playRound('rock') // returns 'player' if we win...
+rps.playerScore       // ...and our score would have increased
+
+// We also have the ability to check the winner and reset the game at any time
+rps.getWinningPlayer() // 'player', if we won above round
+rps.reset()
+```
+
+You may be looking at this code and thinking that you personally prefer to split your code between more functions than you see here, but also recognize that those functions may not really be a useful interaction point for anyone using your object.
+
+But, there is no rule saying that you can't add those functions to your object as well! A common convention is to prefix methods and properties that you don't intend other people to use with an underscore (`_`). This convention conveys to others that "These things are meant to be used internally by this object, please interact with the other available methods and properties on this object's interface instead".
```

Comment:
My thoughts on the `_` bit - I think there's value in introducing the concept of private properties but I don't feel like we need to dive into it at all. 

My suggestion would be to start by introducing the concept of private properties first (as opposed to after the code example), mention that we'll expand on more ways to implement private properties later, then mention that object literal syntax can't implement truly private properties but that the `_` convention existed as "pseudo-private". Again, focus on how we'll go into more ways to implement actually private properties later.

The code example for `_` conventions can also do with the same treatment as the above - I don't think we need a full working RPS example as that might be a red herring taking away from the real concept to learn, and putting effort more into parsing the RPS code.

That'll also significantly shorten this section, which I believe will have a similar effect on focus on concepts throughout the lesson, as it currently feels like a section much larger than it needs to be relative to everything else.

Reply:
> My suggestion would be to start by introducing the concept of private properties first (as opposed to after the code example), mention that we'll expand on more ways to implement private properties later...

I think we can do this with very slight changes, what do you think?
```markdown
### Private Properties

You may be looking at this code and thinking that you personally prefer to split your code between more functions than you see here, but also recognize that those functions may not really be a useful interaction point for anyone using your object.

But, there is no rule saying that you can't add those functions to your object as well! A common convention is to prefix methods and properties that you don't intend other people to use with an underscore (`_`). This convention conveys to others that "These things are meant to be used internally by this object, please interact with the other available methods and properties on this object's interface instead"

Another name for properties like this might be **private properties**/**private methods**, and even though object literal syntax doesn't provide a way to truly prevent people from using them, you will later learn about other methods of creating objects that *can*. 

Private properties/methods aren't strictly required, but they can help make the intended use of the object more understandable, and when used thoughtfully, even protect certain properties (like the player's scores) from being modified in ways that you may not have intended. Back off, cheaters!

`code example, with pseudo-coded method implementations`
```

> ... then mention that object literal syntax can't implement truly private properties but that the _ convention existed as "pseudo-private".

I'll also add a lesson note, or some paragraph in the middle there to draw attention to this. 

Note that I think in order for this section to be effective, we need to decide on how to proceed with the previous comment about the fully working RPS code.

Reply:
Continuing/combining in https://github.com/TheOdinProject/curriculum/pull/27381#discussion_r1807309806